### PR TITLE
feat(react): persist interactive-component savedUserSettings per-user via UserInfoEngine

### PR DESCRIPTION
**Pull request by AN-BC**
## Summary

Interactive (Skip/React) components already receive a `savedUserSettings` prop (in) and an `onSaveUserSettings` callback (out) — but nothing ever persisted them: the read side was never seeded and the write side only emitted a `userSettingsChanged` event that **no code in the repo consumes**. This PR wires that **existing contract** to durable, **per-user, cross-device** persistence via `UserInfoEngine`, automatically scoped per component.

**No change to the component contract.** Components keep the clean single-object-in / event-out pattern they already use; they simply gain persistence for free.

## How it works

- The Angular host (`MJReactComponent`) is the single chokepoint every Angular host path flows through.
  - **On load:** seeds `savedUserSettings` from `UserInfoEngine` (key `InteractiveComponents_UserState_Root/<scope>`), merging stored values over any host-provided defaults (stored wins).
  - **On save:** **merges** the payload over the saved snapshot (see "Merge semantics" below) and persists via `SetSettingDebounced` (debounced, cross-device).
  - Still emits `userSettingsChanged` for back-compat — carrying the merged snapshot, so observers and storage agree.
- **Scope** defaults to `<namespace>/<name>` from the component spec (lowercased), overridable via a new `UserStateScope` input for hosts that render the same spec in multiple distinct contexts.
- New `PersistUserSettings` input (default `true`) lets a host opt out and own persistence itself.

## Merge semantics (from review feedback)

Per @jordanfanapour's review: the host **merges each `onSaveUserSettings` payload over the saved settings instead of replacing wholesale** (`applyUserSettingsUpdate` in `react-runtime`). This protects against two silent data-loss cases:

1. A component (especially AI-generated) passing only the changed keys instead of spreading the full object.
2. The stale-prop case: the host deliberately never re-renders on save, so the `savedUserSettings` prop a component spreads is frozen at mount — under full-replace, a second same-session save would wipe the first change even for a component following the documented pattern.

Removing a key requires explicit intent: pass it with value `null` (or `undefined`) and it's deleted from the stored object; reads fall back to defaults via the documented `?? fallback` pattern. Docs and AI-generator prompt templates still recommend passing the full spread — host-side merge is the safety net, not the contract. The test harness mirrors the same semantics.

## Changes by package

| Package | Change |
|---|---|
| `@memberjunction/react-runtime` | New framework-agnostic `user-state` helpers (`resolveUserStateScope`, `userStateStorageKey`, `parseStoredUserSettings`, `mergeUserSettings`, `applyUserSettingsUpdate`) + unit tests. No persistence logic here — kept layer-clean. |
| `@memberjunction/ng-react` | `MJReactComponent` seeds + persists via `UserInfoEngine` with merge-on-save semantics; new `UserStateScope` / `PersistUserSettings` inputs; provider threaded via `ProviderToUse`. |
| `@memberjunction/react-test-harness` | `ComponentExecutionOptions.savedUserSettings` seeds the prop; `onSaveUserSettings` merges into a real in-memory snapshot (null-removes-key, matching the production host) for assertions. |
| `metadata/components/*`, `metadata/prompts/templates/*` | Teach AI component generators that these props persist per-user/cross-device, that the host merges (partial objects can't wipe other keys), and the explicit-`null` removal convention. |

## Design notes

- Storage is **one blob per component scope** (`InteractiveComponents_UserState_Root/<scope>` → JSON object) — single source of truth, no parent/child lost-update races, and it unifies cleanly with the existing `savedUserSettings` object.
- We deliberately did **not** add a per-key `utilities.settings.Get/Set` sugar API — the existing single-object contract is self-contained and clean; the gap was only persistence.

## Backward compatibility

Fully additive. Confirmed via codebase search: no template binds `[savedUserSettings]` and nothing consumes `userSettingsChanged` today, so enabling auto-persist is safe (no double-write). Existing components using the blob gain durable persistence with zero changes.

## Testing

- `@memberjunction/react-runtime`: **66 passed** (24 new `user-state` tests, incl. delta payloads, stale-prop double-save, null-removes-key).
- `@memberjunction/ng-react`: **14 passed**.
- `@memberjunction/react-test-harness`: **329 passed**, 110 skipped.
- All three packages build cleanly.

Changeset included (patch bump).

https://claude.ai/code/session_017fjNHH44a5JmohUqM2mLCf

---
_Generated by [Claude Code](https://claude.ai/code/session_017fjNHH44a5JmohUqM2mLCf)_

### CODE REVIEW

**Review by AN-BC — COMMENTED** [review]
LGTM

**Review by AN-BC — COMMENTED** [review]
LGTM

**Review by AN-BC — COMMENTED** [review]
LGTM. @jordanfanapour  can you give this a quick look and if LG please merge. This will make Skip component user settings start actually saving! @CaelebB-BC FYI.

File: plans/interactive-component-user-settings-persistence.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
+# Interactive Component — Per-User Settings Persistence
```
Comment (jordanfanapour) [inline plans/interactive-component-user-settings-persistence.md]: This plan document can be removed to keep the directory clean.

File: metadata/components/claude.md [hunk 2]
```diff
@@ -183,8 +183,54 @@ Root components receive these standard props:
...
+
+```javascript
+function MainComponent({ savedUserSettings, onSaveUserSettings, /* ...standard props */ }) {
+  const [sortBy, setSortBy] = React.useState(savedUserSettings?.sortBy ?? 'CloseDate');
+  const [viewMode, setViewMode] = React.useState(savedUserSettings?.viewMode ?? 'grid');
+
+  const handleSort = (col) => {
+    setSortBy(col);
+    // Persist the FULL settings object — the host saves & scopes it per user.
+    onSaveUserSettings?.({ ...savedUserSettings, sortBy: col });
```
Comment (jordanfanapour) [inline metadata/components/claude.md:219]: I'm concerned that AI agents will forget the spread pattern and just pass the delta:

// What an agent might write (wrong):
onSaveUserSettings?.({ sortBy: col });

// Instead of the required pattern:
onSaveUserSettings?.({ ...savedUserSettings, sortBy: col });

If the agent forgets the spread, the full-replace semantics in the host will wipe out every other saved preference. A defensive approach would be to merge in the host instead of full-replace. The component should pass the full blob, but the host would be resilient to deltas:

// In the host's onSaveUserSettings handler:
Object.assign(userSettingsState, settings || {});

// No delete-all-keys step — just overlay
That way if an agent passes only { sortBy: 'Name' }, the existing viewMode: 'grid' survives. And if a well-behaved component passes the full spread, it still works correctly since it's a superset.

The only downside: if a component intentionally wants to remove a key, a simple merge can't express that. But that's a rare edge case — and you could handle it with an explicit null convention ({ removedKey: null }) if needed.

Comment (AN-BC) [reply metadata/components/claude.md:219]: the merge concept makes sense to me as the default behavior and the explicit null is easier for us to make sure the agent doesn't do by mistake. I'll get @claude  to fix this.

Comment (AN-BC) [reply metadata/components/claude.md:219]: Implemented in d8a7ce28a (supersedes the inline version in 27c4def3a — same semantics, factored into the shared tested helper).

The host now merges every `onSaveUserSettings` payload over the saved snapshot via a new pure `applyUserSettingsUpdate` helper in `@memberjunction/react-runtime`; an explicit `null` (or `undefined`) value removes the key. Merge turns out to fix a second bug beyond the forgotten-spread case: the host deliberately never re-renders on save, so the `savedUserSettings` prop a component spreads is frozen at mount — under full-replace, two sequential saves in one session would lose the first change even for a well-behaved component following the documented pattern.

Also in the commit: the test harness stub now uses the same merge + null-removes-key semantics (it was delete-all-then-assign), the merged snapshot is what gets emitted on `userSettingsChanged` so observers and storage agree, 8 new unit tests (delta payload, stale-prop double-save, null removal), and the docs/prompt templates now state the host merges, keep recommending the full spread, and document the explicit-null removal convention. Plan doc removed per the other thread.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by AN-BC** [comment]
@jordanfanapour — both of your review comments are addressed as of `d8a7ce28a`:

1. **Plan doc** — `plans/interactive-component-user-settings-persistence.md` removed.
2. **Merge instead of full-replace** — implemented exactly as you suggested, including the explicit-`null`-removes-key convention. The host now overlays each `onSaveUserSettings` payload onto the saved snapshot via a new pure `applyUserSettingsUpdate` helper in `@memberjunction/react-runtime`, so a delta-only payload from a forgetful agent can't wipe other preferences. It also turned out to fix a second lost-update case: the host intentionally never re-renders on save, so the `savedUserSettings` prop a component spreads is frozen at mount — under full-replace, two sequential saves in one session would lose the first change even for a well-behaved component. The test harness mirrors the same merge semantics, there are unit tests for the delta / stale-prop / null-removal cases, and the docs + AI-generator prompt templates now state that the host merges, keep recommending the full spread, and document `null`-to-clear.

Details in the thread reply on your comment; PR description is updated to match. Ready for re-review.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by AN-BC** [comment]
@jordanfanapour if you're happy with these changes, merge please after reviewing

**Review by AN-BC — COMMENTED** [review]
no changes on my end

**Comment by AN-BC** [comment]
went ahead and merged @jordanfanapour  as I needed this for something else
